In [11]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns


In [12]:
df = pd.read_csv("gym_churn_clean.csv")
df.head()


gender  near_location  partner  promo_friends  phone  contract_period  \
0       1              1        1              1      0                6   
1       0              1        0              0      1               12   
2       0              1        1              0      1                1   
3       0              1        1              1      1               12   
4       1              1        1              1      1                1   

   group_visits  age  avg_aditional_charges_total  month_to_end_contract  \
0             1   29                    14.227470                      5   
1             1   31                   113.202938                     12   
2             0   28                   129.448479                      1   
3             1   33                    62.669863                     12   
4             0   26                   198.362265                      1   

   lifetime  avg_class_freq_total  avg_class_freq_current_month  churn  
0         3              0.020398                      0.000000      0  
1         7              1.922936                      1.910244      0  
2         2              1.859098                      1.736502      0  
3         2              3.205633                      3.357215      0  
4         3              1.113884                      1.120078      0

Agregar columna de diferencia de promedio y frecuencia del ultimo mes y otra de cargos promedio por mes de permnencia, estas frecuencias ayudarán a que pueda haber una mejor predicción 

In [13]:
df["freq_drop"] = df["avg_class_freq_total"] - df["avg_class_freq_current_month"]

df

gender  near_location  partner  promo_friends  phone  contract_period  \
0          1              1        1              1      0                6   
1          0              1        0              0      1               12   
2          0              1        1              0      1                1   
3          0              1        1              1      1               12   
4          1              1        1              1      1                1   
...      ...            ...      ...            ...    ...              ...   
3995       1              1        1              0      1               12   
3996       0              1        0              0      1                1   
3997       1              1        1              1      1               12   
3998       0              1        1              1      1                6   
3999       1              0        1              0      0               12   

      group_visits  age  avg_aditional_charges_total  month_to_end_contract  \
0                1   29                    14.227470                      5   
1                1   31                   113.202938                     12   
2                0   28                   129.448479                      1   
3                1   33                    62.669863                     12   
4                0   26                   198.362265                      1   
...            ...  ...                          ...                    ...   
3995             0   33                     2.406023                     12   
3996             1   29                    68.883764                      1   
3997             0   28                    78.250542                     11   
3998             0   32                    61.912657                      5   
3999             1   30                   178.861814                     12   

      lifetime  avg_class_freq_total  avg_class_freq_current_month  churn  \
0            3              0.020398                      0.000000      0   
1            7              1.922936                      1.910244      0   
2            2              1.859098                      1.736502      0   
3            2              3.205633                      3.357215      0   
4            3              1.113884                      1.120078      0   
...        ...                   ...                           ...    ...   
3995         8              2.284497                      2.349070      0   
3996         1              1.277168                      0.292859      1   
3997         2              2.786146                      2.831439      0   
3998         3              1.630108                      1.596237      0   
3999         1              1.101832                      1.035270      0   

      freq_drop  
0      0.020398  
1      0.012693  
2      0.122596  
3     -0.151582  
4     -0.006194  
...         ...  
3995  -0.064572  
3996   0.984309  
3997  -0.045293  
3998   0.033872  
3999   0.066562  

[4000 rows x 15 columns]

In [14]:
df["monthly_extra_charge"] = df["avg_aditional_charges_total"] / df["lifetime"].replace(0, 1)
df

gender  near_location  partner  promo_friends  phone  contract_period  \
0          1              1        1              1      0                6   
1          0              1        0              0      1               12   
2          0              1        1              0      1                1   
3          0              1        1              1      1               12   
4          1              1        1              1      1                1   
...      ...            ...      ...            ...    ...              ...   
3995       1              1        1              0      1               12   
3996       0              1        0              0      1                1   
3997       1              1        1              1      1               12   
3998       0              1        1              1      1                6   
3999       1              0        1              0      0               12   

      group_visits  age  avg_aditional_charges_total  month_to_end_contract  \
0                1   29                    14.227470                      5   
1                1   31                   113.202938                     12   
2                0   28                   129.448479                      1   
3                1   33                    62.669863                     12   
4                0   26                   198.362265                      1   
...            ...  ...                          ...                    ...   
3995             0   33                     2.406023                     12   
3996             1   29                    68.883764                      1   
3997             0   28                    78.250542                     11   
3998             0   32                    61.912657                      5   
3999             1   30                   178.861814                     12   

      lifetime  avg_class_freq_total  avg_class_freq_current_month  churn  \
0            3              0.020398                      0.000000      0   
1            7              1.922936                      1.910244      0   
2            2              1.859098                      1.736502      0   
3            2              3.205633                      3.357215      0   
4            3              1.113884                      1.120078      0   
...        ...                   ...                           ...    ...   
3995         8              2.284497                      2.349070      0   
3996         1              1.277168                      0.292859      1   
3997         2              2.786146                      2.831439      0   
3998         3              1.630108                      1.596237      0   
3999         1              1.101832                      1.035270      0   

      freq_drop  monthly_extra_charge  
0      0.020398              4.742490  
1      0.012693             16.171848  
2      0.122596             64.724240  
3     -0.151582             31.334932  
4     -0.006194             66.120755  
...         ...                   ...  
3995  -0.064572              0.300753  
3996   0.984309             68.883764  
3997  -0.045293             39.125271  
3998   0.033872             20.637552  
3999   0.066562            178.861814  

[4000 rows x 16 columns]

In [15]:
df.describe()

gender  near_location      partner  promo_friends        phone  \
count  4000.000000    4000.000000  4000.000000    4000.000000  4000.000000   
mean      0.510250       0.845250     0.486750       0.308500     0.903500   
std       0.499957       0.361711     0.499887       0.461932     0.295313   
min       0.000000       0.000000     0.000000       0.000000     0.000000   
25%       0.000000       1.000000     0.000000       0.000000     1.000000   
50%       1.000000       1.000000     0.000000       0.000000     1.000000   
75%       1.000000       1.000000     1.000000       1.000000     1.000000   
max       1.000000       1.000000     1.000000       1.000000     1.000000   

       contract_period  group_visits          age  \
count      4000.000000   4000.000000  4000.000000   
mean          4.681250      0.412250    29.184250   
std           4.549706      0.492301     3.258367   
min           1.000000      0.000000    18.000000   
25%           1.000000      0.000000    27.000000   
50%           1.000000      0.000000    29.000000   
75%           6.000000      1.000000    31.000000   
max          12.000000      1.000000    41.000000   

       avg_aditional_charges_total  month_to_end_contract     lifetime  \
count                  4000.000000            4000.000000  4000.000000   
mean                    146.943728               4.322750     3.724750   
std                      96.355602               4.191297     3.749267   
min                       0.148205               1.000000     0.000000   
25%                      68.868830               1.000000     1.000000   
50%                     136.220159               1.000000     3.000000   
75%                     210.949625               6.000000     5.000000   
max                     552.590740              12.000000    31.000000   

       avg_class_freq_total  avg_class_freq_current_month        churn  \
count           4000.000000                   4000.000000  4000.000000   
mean               1.879020                      1.767052     0.265250   
std                0.972245                      1.052906     0.441521   
min                0.000000                      0.000000     0.000000   
25%                1.180875                      0.963003     0.000000   
50%                1.832768                      1.719574     0.000000   
75%                2.536078                      2.510336     1.000000   
max                6.023668                      6.146783     1.000000   

         freq_drop  monthly_extra_charge  
count  4000.000000           4000.000000  
mean      0.111968             71.908206  
std       0.321350             73.970838  
min      -1.060218              0.035944  
25%      -0.056744             19.149330  
50%       0.022433             46.298186  
75%       0.141110            100.312150  
max       1.856764            552.590740

Resumen inicial: Cantidad de clientes: 

-4000 registros analizados.

-Tasa de cancelación (churn): Aproximadamente 27% de los clientes cancelaron su membresía.

-51% son mujeres.

-85% viven cerca del gimnasio.

-49% tienen pareja.

-31% usaron promociones de amigos.

-90% proporcionaron teléfono.

-41% participan en visitas grupales.

-Edad: Promedio de 29 años, con rango de 18 a 41 años, indicando una clientela mayormente joven-adulta.

-Promedio total: 1.88 clases/mes.

-Promedio último mes: 1.77 clases/mes.

-Diferencia promedio (freq_drop): 0.11, indicando una ligera reducción en la asistencia reciente.

-Existen cargos adicionales:
Promedio total: $147 con alta variabilidad.

Cargos promedio mensuales (monthly_extra_charge): $72 en promedio, con algunos clientes con gastos muy altos.

-Duración del contrato: Promedio de casi 5 meses, con un rango de 1 a 12 meses.

-Duración como cliente (lifetime): Promedio 3.7 meses, con valores hasta 31 meses.